docs/design-docs/03-pipeline-merge.md

```diff
@@ -0,0 +1,628 @@
+Parameters(setters) are the primary interface for the package. This is an anti-pattern for package-as-interface motivation 
+and one of the major blockers for not using other functions is the merge behavior 
+of the pipeline. If this problem is solved, LZ maintainers can rewrite the packages 
+with best practices aligned to the bigger goal of treating the package-as-interface, 
+discourage excessive use of setters and only use them as parameterization techniques of last resort.
+
+## Design
+
+In order to solve this problem, we should merge the pipeline section of the Kptfile 
+in a custom manner based on the most common use-cases and the expected outputs in 
+such scenarios. There are no user interface changes. Users will invoke the `kpt pkg 
+update` command in the same way they do currently. This effort will improve the merged 
+output of the pipeline section.
+
+**Is this change backwards compatible**: We are not making any changes to the api, 
+we are only improving the merged output of the `pipeline` section. This change will 
+produce a different output of Kptfile when compared to the current version but this 
+is not a breaking change.
+
+## User guide
+
+Here is what users can expect when they invoke the command, `kpt pkg update` on a package.
+
+#### Terminology
+
+**Original**: Source of the last fetched version of the package, represented by the upstreamLock section in Kptfile.
+
+**Updated upstream**: Declared source of the updated package, represented by the upstream section in Kptfile.
+
+**Local**: Local fork of the package on disk.
+
+Firstly, we need to define the identity of the function in order to uniquely identify 
+a function across three sources to perform a 3-way merge. In order to reliably identify the instance of a
+function, we should add a new optional field `name` to function definition.
+
+We will introduce a new field `name` for identifying functions and merge pipeline as associative list.
+The aim is to encourage users eventually to have `name` field specified for all functions
+similar to containers in deployment. But in the meanwhile, we will be using image name in 
+order to identify the function and make it an associative list for merging. The limitation
+of this approach is the multiple functions with same image can be declared in the pipeline.
+In that case we will fall back to current merge logic of merging pipeline as non-associative list.
+In order to have deterministic behavior, we will use image as merge key iff none of the 
+functions have `name` field specified and iff there are no duplicate image names declare in functions.
+
+So the algorithm to merge:
+
+1. If `name` field is specified in all the functions across all sources, merge 
+it as associative list using `name` field as merge key.
+2. If `name` field is not specified in all the functions across all sources,
+and if there is no duplicate declaration of image names for functions, use `image` value
```

Comment:
when we say `image`, do we mean `image-name` with `version` or just the `image` ?  for ex.

```
pipeline:
  mutators:
     - image: gcr.io/kpt-fn/set-namespace:v0.1.1
```

Is the key `gcr.io/kpt-fn/set-namespace:v0.1.1` or `gcr.io/kpt-fn/set-namespace` ?

Reply:
No `version`. Please see next line.